### PR TITLE
Fix hard rollbacks for TM 

### PR DESCRIPTION
**Pull request by BrandonWeng**
## Describe your changes and provide context
The block app hash is the committed app state from the previous height, therefore the latest block tracks the app hash of the N - 1. 

If we're deleting block N, then we need to rollback the tendermint state to N - 1, this way the block N can be applied again 

## Testing performed to validate your change
Testing here https://github.com/sei-protocol/sei-cosmos/pull/119

### CODE REVIEW

File: internal/state/rollback.go [hunk 1]
```diff
@@ -64,18 +81,33 @@ func Rollback(bs BlockStore, ss Store, removeBlock bool, privValidatorConfig *co
...
+	rolledBackLastResultHash := latestBlock.Header.LastResultsHash
+
+	// Only the latest state is stored
+	fmt.Printf("Latest state Height=%d, appHash=%X\n", latestState.LastBlockHeight, latestState.AppHash)
+
+	fmt.Printf("Rollback block Height=%d, appHash=%X\n", rollbackBlock.Header.Height, rollbackBlock.Header.AppHash)
+	fmt.Printf("Rollback block Height=%d, lastCommitHash=%X\n", rollbackBlock.Header.Height, rollbackBlock.Header.LastCommitHash)
+	fmt.Printf("Latest block Height=%d, appHash=%X\n", latestBlock.Header.Height, latestBlock.Header.AppHash)
+	fmt.Printf("latest block Height=%d, lastCommitHash=%X\n", latestBlock.Header.Height, latestBlock.Header.LastCommitHash)
+
```
Comment (BrandonWeng) [inline internal/state/rollback.go:109]: Likewise here, just being more verbose around the state that we're modifying

File: internal/store/store.go [hunk 2]
```diff
@@ -778,5 +778,9 @@ func (bs *BlockStore) DeleteLatestBlock() error {
 	if err != nil {
 		return fmt.Errorf("failed to delete height %v: %w", targetHeight, err)
 	}
+
+	if err := batch.Close(); err != nil {
+		panic(err)
```
Comment (philipsu522) [inline internal/store/store.go:783]: does it matter if we panic here?

**Review by philipsu522 — APPROVED** [review]

File: internal/state/rollback.go [hunk 3]
```diff
@@ -86,23 +118,23 @@ func Rollback(bs BlockStore, ss Store, removeBlock bool, privValidatorConfig *co
...
+		NextValidators:              latestState.Validators,
+		Validators:                  latestState.LastValidators,
 		LastValidators:              previousLastValidatorSet,
 		LastHeightValidatorsChanged: valChangeHeight,
 
 		ConsensusParams:                  previousParams,
 		LastHeightConsensusParamsChanged: paramsChangeHeight,
-
-		LastResultsHash: latestBlock.Header.LastResultsHash,
-		AppHash:         latestBlock.Header.AppHash,
```
Comment (philipsu522) [inline internal/state/rollback.go:105]: This is the main issue right? Before your PR we were using block N's hashes, while in this PR we use N-1's hashes

File: internal/store/store.go (hunk 2 above)
Comment (BrandonWeng) [reply internal/store/store.go:783]: it's what OSS is doing rightn ow

File: internal/state/rollback.go (hunk 3 above)
Comment (BrandonWeng) [reply internal/state/rollback.go:105]: yeah

**Pull request merged**